flybywiresim/a32nx
# fix(nd): TCAS symbols correctly display in all nav/rose views, fixed TCAS related TypeError exception to console

Fixes #4952
Fixes #4581

Note: This is not a complete TCAS overhaul, this is just to fix some semi-critical and long standing issues with the default Asobo TCAS ND display. A proper implementation of ADS-B / ATSAW (#5014) is coming. 

## Summary of Changes


- [x] Reproduce and fix TCAS related altText.textcontent TypeError exception - Caused by updateDraw() being called before createDraw() and trying to look up variables that have not been init
- [x] TCAS symbols drawn on the 40/80/160/320 ARC views
- [x] TCAS symbols drawn on the 40/80/160/320 nm ROSE view

This PR fixes two bugs, 

Firstly a frustrating and hard to reproduce bug where the JS console would be flooded with stack trace errors whenever an AI plane spawned in, sometimes causing the TCAS symbol for that plane to bug out. It's not entirely known how to reproduce this error, but ensuring that a certain bit of code is called **before** the problem causing code seems to solve the issue. 

Secondly a bug where the 40 nm ROSE/ARC view was not displaying TCAS symbols, because of the maximum TCAS NPC aircraft symbol display range being incorrectly processed.

## Screenshots (if necessary)



1. altText bug with error output to js console (Left) vs Fixed (Right)

![FlightSimulator_JLqQXV3QCh](https://user-images.githubusercontent.com/15316958/121192134-346f5980-c89f-11eb-91bf-09fe845f17df.jpg)

2. 40 nm ROSE/NAV view now displays TCAS symbols

![FlightSimulator_0oDvF4HRcu](https://user-images.githubusercontent.com/15316958/121192262-58cb3600-c89f-11eb-9c45-735c512ce947.jpg)

![v85idg4Js2](https://user-images.githubusercontent.com/15316958/121192299-6385cb00-c89f-11eb-9078-42a0db20cb40.png)

(TCAS symbols display up to ~30 nm, this is correct behavior according to documentation and intended behaviour)

![xu3narzuTD](https://user-images.githubusercontent.com/15316958/121192333-6bde0600-c89f-11eb-9515-045857ee1eee.png)


## References


> G. Information Display The TCAS information is presented on the CAPT and F/O NDs of the EFIS
 system. Additional messages can also be presented on the display units ofthe ECAM system.
> (1) Traffic information on the ND
> Target aircraft are presented on the ND in ROSE or ARC mode but not in PLAN mode. Traffic indications show the situation in the zone. Aircraft present in this zone are represented by symbols whose shape and color correspond to the type of intruders defined in the TCAS. 
> ...
These indications are only presented for the 10, 20 and 40 NM selections. 
> ...

>TCAS intruders shall be displayed on ND Capt and F/O in all Rose modes and ARC mode:
>Full time option installed: For all ranges except in ZOOM ranges, NDRange > 9NM,
>Full time option not installed: For all ranges <=41 nm except ZOOM ranges: NDRANGE > 9nm
https://www.aviationmatters.co/tcas-guide/
http://a320dp.com/A320_DP/warning-systems/sys-16.3.2.html
https://www.icao.int/APAC/Meetings/2011_ADS_B_SITF10/SP07b_Airbus%20-%20ADS-B%20IN.pdf
https://www.aviatorsbuzz.com/techtalks/a320s-traffic-alert-and-collision-avoidance-system-tcas/




## Additional context

Really annoying to debug


Discord username (if different from GitHub): 2Cas#1022

## Testing instructions


1. Turn Live Traffic on in the options
2. Use Coherent GT Debugger (SDK) to observe the JS console (A320_Neo_MFD_1/A320_Neo_MFD_2)
3. Spawn in at a major airport with lots of AI traffic planes around. (i.e. Charles De Gaulle airport, JFK, etc.)
4. Start your plane as normal, turn TCAS to TA/RA 
5. Either fly a normal flight and try to stay within TCAS range of surrounding traffic, or use slew mode to position your aircraft facing a runway and adjusting altitude by slewing up/down to deliberately have TCAS intruding traffic.
6. Turn to the 40 nm NAV/ROSE views, these should show TCAS traffic symbols when intercepted by traffic. 
7. Check that the 20nm/10nm NAV/ROSE views still work as expected
8. Check that 80/160/320 views still do not show TCAS traffic
9. No exceptions should print to the console. Note: As this bug is hard to reproduce, around 10-15 minutes of traffic should pass without any TypeError javascript exceptions printing to the console. 
10. Turn TCAS to STDBY and check that this also does not cause errors to be printed to the console

Keep an eye out on regular flights outside of dedicated testing when you encounter traffic that the ND displays TCAS symbols as expected.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
